sklearn/decomposition/incremental_pca.py

```diff
@@ -198,7 +202,10 @@ def fit(self, X, y=None):
 
         for batch in gen_batches(n_samples, self.batch_size_,
                                  min_batch_size=self.n_components or 0):
-            self.partial_fit(X[batch], check_input=False)
+            X_batch = X[batch]
+            if sparse.issparse(X_batch):
```

Comment:
Maybe we want the densification to happen in `partial_fit`, so that `partial_fit` also supports sparse matrices?

Reply:
Densification inside `partial_fit` would be the same as densifying before passing to `partial_fit`, which I feel would be inconsistent with the choice to not densify sparse matrices passed to `decomposition.PCA`. If others agree on this I don't feel strongly, though.

Reply:
If `partial_fit` doesn't densify, then the claim that the estimator works with sparse data is wrong and should be adjusted.

I'm not sure this densifying in `partial_fit `is something we actually want though. thoughts @jnothman ?

Reply:
Hmm... I'd be more comfortable atm not supporting sparse data in `partial_fit`, but supporting sparse in `fit` by densifying each batch. Will that be weird?

Reply:
I could add a nice error catching sparse matrices in `partial_fit` recommending that the user use `fit` instead if that makes it less confusing?

Reply:
Well if they want to train incrementally, and you were otherwise proposing to densify their data, recommending that they do that (i.e. densify) seems more appropriate.

Reply:
Something like `TypeError: IncrementalPCA.partial_fit does not support sparse input. Either convert data to dense or use IncrementalPCA.fit to do so in batches.` fit the bill?

Reply:
Edit: I see what you mean - if someone has _multiple_ sparse matrices that they want to use to fit, they would not be able to do so with what we have written here. Does this call for a third function? I'm not inclined to make the API more complex, though.

To show what I mean, this is the current behaviour of `fit`: 1. Reset object. 2. Apply `partial_fit` batchwise.

If I want to apply `partial_fit` batchwise to multiple input matrices, I would not be able to run step 2 without step 1.

Reply:
hey had multiple matrices, they would need to call fit multiple times anyway, whatever the estimator.

I'm happy with the TypeError message you proposed (replacing the second `fit` with `partial_fit` ;))